### PR TITLE
feat(forge) add debugger breakpoints

### PR DESCRIPTION
**Pull request by iFrostizz**
## Motivation

Adding a new breakpoints feature in order to jump to these positions in the debugger by using vim-like jumplist `'<char>`.

This is an example of the using of this new func:

```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.13;

import "forge-std/Test.sol";
import "../src/Counter.sol";

contract CounterTest is Test {
    Counter public counter;

    function setUp() public {
        counter = new Counter();
    }

    function testIncrement() public {
        vm.breakpoint("a");

        counter.increment();
        assertEq(counter.number(), 1);

        vm.breakpoint("c");
    }
}
```

```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.13;

import "forge-std/Test.sol";

contract Counter is Test {
    uint256 public number;

    function setNumber(uint256 newNumber) public {
        number = newNumber;
    }

    function increment() public {
        number++;
        vm.breakpoint("b");
    }
}
```

Running the tests:

```sh
forge t --debug testIncrement
```

View from the debugger:

image

Hitting `'a`:

image

Hitting `'b`:

image

Hitting `'c`:

image

Closes https://github.com/foundry-rs/foundry/issues/3430

## Solution

Solution was to implement the new `breakpoint` cheatcode, store the PC as well as the caller (contract) of the current context. It's then filtered in the debugger until finding the matching address & PC in the debug steps.

### CODE REVIEW

File: ui/src/lib.rs [hunk 1]
```diff
@@ -410,17 +413,19 @@ impl Tui {
...
                                         .split_inclusive('\n')
                                         .collect::<Vec<&str>>();
-                                    let actual = source[offset..offset + len]
+                                    let actual = source[std::cmp::min(offset, max)..
+                                        std::cmp::min(offset + len, max)]
                                         .split_inclusive('\n')
                                         .map(|s| s.to_string())
                                         .collect::<Vec<String>>();
-                                    let mut after = source[offset + len..]
+                                    let mut after = source[std::cmp::min(offset + len, max)..]
```
Comment (iFrostizz) [inline ui/src/lib.rs:428]: fixed out of bounds accesses in the Solidity source code debugger view

**Comment by iFrostizz** [comment]
any progress @mattsse @gakonst ?

**Comment by gakonst** [comment]
This is pretty cool! I'm curious what @brockelmore and @mds1 think here in terms of UX.

File: cli/src/cmd/forge/debug.rs [hunk 2]
```diff
@@ -41,7 +42,7 @@ pub struct DebugArgs {
 }
 
 impl DebugArgs {
-    pub async fn debug(self) -> eyre::Result<()> {
+    pub async fn debug(self, breakpoints: HashMap<char, (Address, usize)>) -> eyre::Result<()> {
```
Comment (mattsse) [inline cli/src/cmd/forge/debug.rs:45]: needs docs what this is,
I'd prefer a wrapper type for the breakpoints arg

File: evm/src/executor/inspector/cheatcodes/env.rs [hunk 3]
```diff
@@ -191,6 +191,29 @@ fn get_recorded_logs(state: &mut Cheatcodes) -> Bytes {
...
 }
 
+fn add_breakpoint(state: &mut Cheatcodes, caller: Address, inner: &str) -> Result<Bytes, Bytes> {
+    let mut chars = inner.chars();
+    let point = chars.next();
+
+    if point.is_none() {
+        return Err("Please provide at least one char for the breakpoint".encode().into())
+    };
+    let point = point.unwrap();
```
Comment (mattsse) [inline evm/src/executor/inspector/cheatcodes/env.rs:201]: use unwrap_or_else here

File: ui/src/lib.rs [hunk 4]
```diff
@@ -1021,205 +1026,232 @@ impl Ui for Tui {
...
+                if self.key_buffer.ends_with('\'') {
+                    if let Some((caller, pc)) = self.breakpoints.get(&c) {
+                        for (i, (_caller, debug_steps, _)) in debug_call.iter().enumerate() {
+                            if _caller == caller {
+                                if let Some(step) =
+                                    debug_steps.iter().position(|step| step.pc == *pc)
+                                {
+                                    draw_memory.inner_call_index = i;
+                                    self.current_step = step;
+                                    break
```
Comment (mattsse) [inline ui/src/lib.rs:1042]: what's the change here?
I'd like to see docs for this

File: ui/src/lib.rs [hunk 5]
```diff
@@ -1250,6 +1282,24 @@ enum Interrupt {
...
+                    None
+                }
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    }
+}
```
Comment (mattsse) [inline ui/src/lib.rs:1301]: please simplify by collapsing, use `return Some()` and None as default return end of function

**Review by mattsse — CHANGES_REQUESTED** [review]
sorry for the delay.

This looks useful, but haven't tested it tbh.

left a few nits, My concern is that I'm not very familiar with the debugger changes so it would be very hard for me to debug issues with his, so I'd appreciate thorough docs for the new keycode commands

File: evm/src/executor/inspector/cheatcodes/env.rs (hunk 3 above)
Comment (DaniPopes) [reply evm/src/executor/inspector/cheatcodes/env.rs:201]: more like `ok_or("...". ...)?`

File: ui/src/lib.rs (hunk 4 above)
Comment (iFrostizz) [reply ui/src/lib.rs:1042]: Had to scope it to avoid repeating the part about the matching of the pressed char

**Comment by mds1** [comment]
Thanks @iFrostizz! This is pretty cool. Two comments:
1. Can we add a bool overload, i.e. `vm.breakpoint(string, bool)`? This way conditional breakpoints are supported, which is really helpful when a line is executed many times during a test, but you only want to stop on e.g. the last execution
2. I'm not able to get the `'a` shortcut to jump to my breakpoint? In my test I have `vm.breakpoint("a");` and pressing "apostrophe then a" OR "apostrophe and a" both don't seem to work. I think this might be a vim shortcut and I don't use vim so maybe I'm doing it wrong? We should also add instructions in the debugger, similar to how it says e.g. `[k/j]: prev/next op`

**Comment by mds1** [comment]
> Let's do it, should we make another breakpoint cheatcode, one with only the char and the other with the bool added ? This would be less verbose and error prone that way probably.

Yes sorry that is exactly what I meant! Two cheatcodes, one that just takes `string` and is always active, one that takes `(string, bool)` and is only active when `bool` is true. This overload can also be done in a separate PR if you prefer.

> Weird, are you using the terminal buffer by any chance ? Not sure why the vim bindings would interfere here, but maybe that it's only for the a ?

It's the same with `b` and other letters. I'm using warp terminal so it could be an issue specific to that, I'll try another terminal later but if it's working for everyone else than it's probably an issue on my end

**Comment by iFrostizz** [comment]
@mds1 I'm supportive of both options, what do you think @mattsse ?

**Comment by iFrostizz** [comment]
Actually the change would be less than 20 lines probably, adding the cheatcode and the logic to remove the breakpoint from the HashMap when it's false and add it when it's true

**Comment by iFrostizz** [comment]
@mds1 in case you wrote the breakpoint in the setUp, it seems like they don't work here. Just got the issue.

**Comment by mds1** [comment]
Ok, confirmed it's working as expected now! Not sure what I did wrong last time, but this is awesome 🔥

**Comment by iFrostizz** [comment]
Thanks for confirming @mds1 !
Quick q: do you like where the breakpoint lands into the debugger ? It currently goes on the "call" to the cheatcode, but maybe that it would be better to target next opcodes

**Comment by mds1** [comment]
UX-wise I’d expect the breakpoint cheat call line to be the highlighted line when jumping, and it was, so I’m ok with the current behavior. I’m also ok with tweaking which opcode it lands on as long as the highlighted line remains the same. There’s arguments for jumping to either the first or last opcode that’s part of the breakpoint call, so I’m indifferent here

**Review by mattsse — APPROVED** [review]
alright, this is fantastic, ty for being patient with this PR

lgtm

> should we make the changes in this PR or another one ?

yes, that'd be great!

**Comment by mds1** [comment]
@iFrostizz awesome work here!! 🙌 can you also open a PR to forge-std to add the cheatcode to `Vm.sol`?

**Comment by iFrostizz** [comment]
> @iFrostizz awesome work here!! 🙌 can you also open a PR to forge-std to add the cheatcode to `Vm.sol`?

My bad ! I'll do that in a minute

**Pull request merged**